gpflow/posterior.py

```diff
@@ -235,31 +512,67 @@ def predict_f(
+        Kmm = covariances.Kuu(
+            self.inducing_variable, self.kernel, jitter=default_jitter()
+        )  # [M, L, M, L]
+        Kmn = covariances.Kuf(self.inducing_variable, self.kernel, Xnew)  # [M, L, N, P]
+        Knn = self.kernel(
+            Xnew, full_cov=full_cov, full_output_cov=full_output_cov
+        )  # [N, P](x N)x P  or  [N, P](x P)
 
-    # TODO: this assumes that Xnew has shape [N, D] and no leading dims
+        M, L, N, K = tf.unstack(tf.shape(Kmn), num=Kmn.shape.ndims, axis=0)
+        Kmm = tf.reshape(Kmm, (M * L, M * L))
 
-    Kuf = covariances.Kuf(inducing_variable, kernel, Xnew)  # [(R), M, N]
+        if full_cov == full_output_cov:
+            Kmn = tf.reshape(Kmn, (M * L, N * K))
+            Knn = tf.reshape(Knn, (N * K, N * K)) if full_cov else tf.reshape(Knn, (N * K,))
+            mean, cov = base_conditional(
+                Kmn, Kmm, Knn, self.q_mu, full_cov=full_cov, q_sqrt=self.q_sqrt, white=self.whiten
+            )  # [K, 1], [1, K](x NK)
+            mean = tf.reshape(mean, (N, K))
+            cov = tf.reshape(cov, (N, K, N, K) if full_cov else (N, K))
+        else:
+            Kmn = tf.reshape(Kmn, (M * L, N, K))
+            mean, cov = fully_correlated_conditional(
+                Kmn,
+                Kmm,
+                Knn,
+                self.q_mu,
+                full_cov=full_cov,
+                full_output_cov=full_output_cov,
+                q_sqrt=self.q_sqrt,
+                white=self.whiten,
+            )
+        return mean, cov
 
-    if isinstance(kernel, (kernels.SeparateIndependent, kernels.IndependentLatent)):
-        # NOTE calling kernel(Xnew, full_cov=full_cov, full_output_cov=False) directly would return
-        # if full_cov: [P, N, N] -- this is what we want
-        # else: [N, P] instead of [P, N] as we get from the explicit stack below
-        Knn = tf.stack([k(Xnew, full_cov=full_cov) for k in kernel.kernels], axis=0)
-    elif isinstance(kernel, kernels.MultioutputKernel):
-        # effectively, SharedIndependent path
-        Knn = kernel.kernel(Xnew, full_cov=full_cov)
-        # NOTE calling kernel(Xnew, full_cov=full_cov, full_output_cov=False) directly would return
-        # if full_cov: [P, N, N] instead of [N, N]
-        # else: [N, P] instead of [N]
-    else:
-        # standard ("single-output") kernels
-        Knn = kernel(Xnew, full_cov=full_cov)  # [N, N] if full_cov else [N]
 
-    return Kuf, Knn
+class FallbackIndependentLatentPosterior(FullyCorrelatedPosterior):  # XXX
+    def _conditional_fused(
+        self, Xnew, full_cov: bool = False, full_output_cov: bool = False
+    ) -> MeanAndVariance:
+        Kmm = covariances.Kuu(
+            self.inducing_variable, self.kernel, jitter=default_jitter()
+        )  # [L, M, M]
+        Kmn = covariances.Kuf(self.inducing_variable, self.kernel, Xnew)  # [M, L, N, P]
+        Knn = self.kernel(
+            Xnew, full_cov=full_cov, full_output_cov=full_output_cov
```

Comment:
Should the `full_output_cov` argument be removed?

Reply:
I don't think so; it changes the shape of Knn....